lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -6759,145 +6793,144 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
-    AllocStackInst *subscriptBuffer =
-        builder.createAllocStack(ai->getLoc(), eltType);
-    auto swiftModule = getModule().getSwiftModule();
-    auto diffProto = ctx.getProtocol(KnownProtocolKind::Differentiable);
-    auto diffConf = swiftModule->lookupConformance(astType, diffProto);
+    // %index_int = struct $Int (%index_literal)
+    auto *eltIndexInt = builder.createStruct(loc, intType, {eltIndexLiteral});
+    auto *swiftModule = getModule().getSwiftModule();
+    auto *diffProto = ctx.getProtocol(KnownProtocolKind::Differentiable);
+    auto diffConf = swiftModule->lookupConformance(eltTanAstType, diffProto);
     assert(!diffConf.isInvalid() && "Missing conformance to `Differentiable`");
-    auto addArithProto = ctx.getProtocol(KnownProtocolKind::AdditiveArithmetic);
-    auto addArithConf = swiftModule->lookupConformance(astType, addArithProto);
+    auto *addArithProto =
+        ctx.getProtocol(KnownProtocolKind::AdditiveArithmetic);
+    auto addArithConf =
+        swiftModule->lookupConformance(eltTanAstType, addArithProto);
     assert(!addArithConf.isInvalid() &&
            "Missing conformance to `AdditiveArithmetic`");
-    auto subMap =
-        SubstitutionMap::get(genericSig, {astType}, {addArithConf, diffConf});
-    builder.createApply(ai->getLoc(), fnRef, subMap,
-                        {subscriptBuffer, intStruct, adjointArray});
-    return subscriptBuffer;
-  }
-
-  void accumulateArrayTangentSubscriptDirect(ApplyInst *ai, SILType eltType,
-                                             StoreInst *si,
-                                             AllocStackInst *subscriptBuffer) {
-    auto newAdjValue = builder.emitLoadValueOperation(
-        ai->getLoc(), subscriptBuffer, LoadOwnershipQualifier::Take);
-    recordTemporary(newAdjValue);
+    auto subMap = SubstitutionMap::get(subscriptFnGenSig, {eltTanAstType},
+                                       {addArithConf, diffConf});
+    // %elt_adj = alloc_stack $T.TangentVector
+    auto *eltAdjBuffer = builder.createAllocStack(loc, eltTanType);
+    // apply %subscript_fn<T.TangentVector>(%elt_adj, %index_int, %array_adj)
+    builder.createApply(loc, subscriptFnRef, subMap,
+                        {eltAdjBuffer, eltIndexInt, arrayAdjoint});
+    return eltAdjBuffer;
+  }
+
+  /// Accumulate array element adjoint buffer into `store` source.
+  void accumulateArrayElementAdjointDirect(StoreInst *si,
+                                           AllocStackInst *eltAdjBuffer) {
+    auto eltAdjValue = builder.emitLoadValueOperation(
+        si->getLoc(), eltAdjBuffer, LoadOwnershipQualifier::Take);
+    recordTemporary(eltAdjValue);
     SILValue src = si->getSrc();
     // When the store's source is a `copy_value`, the `copy_value` is part of
     // array literal initialization. In this case, add the adjoint to the source
     // of the copy directly.
     if (auto *cvi = dyn_cast<CopyValueInst>(src))
```

Comment:
Is this necessary for correctness, or is it just an optimization? Could you add this info to the comment?

Reply:
The special `copy_value` source of `store` logic was added in https://github.com/apple/swift/pull/26737

The PR description suggests that the logic is necessary for correctness, but all tests continue to pass if the code is removed. I'll remove the code.

Reply:
Sounds good!

Based on a quick glance at that PR, it looks like that used to be necessary for correctness because of a similar problem to the problem that the current PR is solving. So now that you have solved the problem more generally, this special case is no longer necessary. Yay for simplification.